README.md
# OCC (Orlaco Camera Configurator)

This application can be used to configure some aspects of Orlaco EMOS IP Cameras.

## Background

I was working on a project that required multiple rugged IP cameras and
ended up ordering some Orlaco EMOS IP cameras. These looked ideal for the
application, but I didn't really know much about them in advance.

When they arrived, I found that configuring them required a PC application
that Orlaco themselves don't supply (or if they do, I've not been able to find it).
So, after some research I discovered they use ISO 17215 protocol which sits on top
of SOME/IP protocol, and that configuration utilities are available commercially.
I obtained a quote for one such utility that came in at around 1000 Euros.
Rather a lot I thought, and far more than we wanted to pay, so as any self
respecting software engineer would do, with the help of Google and Wireshark,
I knocked up this utility to allow various settings to be changed.

Note: I don't have access to ISO17215, I've only seen the snippets that are
available for free on the various websites selling it, as such there are some
features that I don't have documentation for. If anyone has
access to a commercial configuration utility and a copy of wireshark, logs
and more info would be very welcome so that more features can be completed.

Hopefully this will be useful to someone, please let me know if so!

## Binaries
I've included a binary built for 64 bit Windows. Linux binaries
are not included, but I have tested that it builds and works under Ubuntu.


## Use
Open a command prompt and type something like this to get a list of command line options:
~~~
occ.exe --help
~~~

## Examples

### Discovering cameras on a network:
~~~
.\occ.exe -d 192.168.2.255  
Found 2 devices  
0: IP=192.168.2.10 Type=01 ServiceID=433f InstanceID=000a V=1.0  
1: IP=192.168.2.11 Type=01 ServiceID=433f InstanceID=000b V=1.0  
~~~
### Read all registers on the camera with IP address 192.168.2.10
~~~
.\occ.exe -R : -i 192.168.2.10  
Read registers 0 to 64

Registers  
Index   Address Hex     Decimal Ascii   Name  
00      0xb00c  0x01      1             LED Mode  
01      0xb041  0x00      0             Stream Protocol  
02      0xb042  0xc0    192             IP Address 0  
03      0xb043  0xa8    168             IP Address 1  
04      0xb044  0x02      2             IP Address 2  
05      0xb045  0x0a     10             IP Address 3  
06      0xb046  0xff    255             Network Mask 0  
07      0xb047  0xff    255             Network Mask 1  
08      0xb048  0xff    255             Network Mask 2  
09      0xb049  0x00      0             Network Mask 3  
10      0xb04a  0x70    112     p       MAC Address 0  
11      0xb04b  0xb3    179             MAC Address 1  
12      0xb04c  0xd5    213             MAC Address 2  
13      0xb04d  0x4f     79     O       MAC Address 3  
14      0xb04e  0x53     83     S       MAC Address 4  
15      0xb04f  0x34     52     4       MAC Address 5  
16      0xb055  0x00      0             VLAN ID 0  
17      0xb056  0x00      0             VLAN ID 1  
18      0xb057  0x45     69     E       Stream ID 0  
19      0xb058  0x4d     77     M       Stream ID 1  
20      0xb059  0x4f     79     O       Stream ID 2  
21      0xb05a  0x53     83     S       Stream ID 3  
22      0xb05b  0x20     32             Stream ID 4  
23      0xb05c  0x20     32             Stream ID 5  
24      0xb05d  0x20     32             Stream ID 6  
25      0xb05e  0x20     32             Stream ID 7  
26      0xb05f  0xc0    192             Destination IP Address 0  
27      0xb060  0xa8    168             Destination IP Address 1  
28      0xb061  0x02      2             Destination IP Address 2  
29      0xb062  0x01      1             Destination IP Address 3  
30      0xb063  0xff    255             Destination MAC Address 0  
31      0xb064  0xff    255             Destination MAC Address 1  
32      0xb065  0xff    255             Destination MAC Address 2  
33      0xb066  0xff    255             Destination MAC Address 3  
34      0xb067  0xff    255             Destination MAC Address 4  
35      0xb068  0xff    255             Destination MAC Address 5  
36      0xb069  0xc3    195             Destination Port 0  
37      0xb06a  0x54     84     T       Destination Port 1  
38      0xb06b  0x01      1             Selected ROI  
39      0xb06c  0x00      0             No Stream At Boot ?  
40      0xb06d  0x43     67     C       UDP Communication Port 0  
41      0xb06e  0x3f     63     ?       UDP Communication Port 1  
42      0xb06f  0xc3    195             RTP Stream Source Port 0  
43      0xb070  0x54     84     T       RTP Stream Source Port 1  
44      0xb071  0x01      1             HDR ?  
45      0xb072  0x01      1             Overlay ?  
46      0xb073  0x00      0             DHCP Enabled ?  
47      0xb078  0x00      0             Wait For MAC ?  
48      0xb079  0x00      0             Wait For PTP Sync ?  
49      0xb171  0x4f     79     O       DHCP Hostname 0  
50      0xb172  0x72    114     r       DHCP Hostname 1  
51      0xb173  0x6c    108     l       DHCP Hostname 2  
52      0xb174  0x61     97     a       DHCP Hostname 3  
53      0xb175  0x63     99     c       DHCP Hostname 4  
54      0xb176  0x6f    111     o       DHCP Hostname 5  
55      0xb177  0x20     32             DHCP Hostname 6  
56      0xb178  0x45     69     E       DHCP Hostname 7  
57      0xb179  0x4d     77     M       DHCP Hostname 8  
58      0xb17a  0x4f     79     O       DHCP Hostname 9  
59      0xb17b  0x53     83     S       DHCP Hostname 10  
60      0xb17c  0x20     32             DHCP Hostname 11  
61      0xb17d  0x20     32             DHCP Hostname 12  
62      0xb17e  0x20     32             DHCP Hostname 13  
63      0xb17f  0x20     32             DHCP Hostname 14  
64      0xb180  0x20     32             DHCP Hostname 15
~~~

### Read the LED Mode register
~~~
.\occ.exe -r 0 -i 192.168.2.10  
Read register 0
  
Registers  
Index   Address Hex     Decimal Ascii   Name  
00      0xb00c  0x01      1             LED Mode  
~~~

### Write the LED Mode register

~~~
.\occ -w 0=0 -i 192.168.2.10  
Write register 0 value 00  

Registers  
Index   Address Hex     Decimal Ascii   Name  
00      0xb00c  0x00      0             LED Mode
~~~

### Read all regions of interest
~~~
 .\occ.exe -G : -i 192.168.2.10
Read ROI's 1 to 10  

Regions Of Interest  
ROI     P1X     P1Y     P2X     P2Y     Width   Height  Mbps    Fps     Mode  
1       0       0       1280    960     1280    960     5       25      2  
2       0       0       1280    960     1280    960     5       25      1  
3       0       960     1280    0       1280    960     50      30      1  
4       0       0       1280    960     800     600     50      30      1  
5       1280    0       0       960     800     600     50      30      1  
6       0       960     1280    0       800     600     50      30      1  
7       0       0       1280    960     640     480     50      30      1  
8       1280    0       0       960     640     480     50      30      1  
9       0       960     1280    0       640     480     50      30      1  
10      1280    960     0       0       1280    960     50      30      1  
~~~

### Configure region of interest 10
~~~
 .\occ.exe -s 10=1280,960,0,0,1280,960,60,25,2 -i 192.168.2.10

Regions Of Interest  
ROI     P1X     P1Y     P2X     P2Y     Width   Height  Mbps    Fps     Mode  
10      1280    960     0       0       1280    960     60      25      2  
~~~


Changelog.txt
December 22nd 2021
	First version released.

common.h
#ifndef COMMON_H
#define COMMON_H

/****************************************************************************/
/***        Include files                                                 ***/
/****************************************************************************/

#include <stdint.h>

/****************************************************************************/
/***        Macro Definitions                                             ***/
/****************************************************************************/

/*--------------------------------------------------------------------------*/
/*          Boolean Constants                                               */
/*--------------------------------------------------------------------------*/

#if !defined FALSE && !defined TRUE
#define TRUE            (1)   /* page 207 K+R 2nd Edition */
#define FALSE           (0)
#endif /* !defined FALSE && #if !defined TRUE */

/*--------------------------------------------------------------------------*/
/*          Null pointers                                                   */
/*--------------------------------------------------------------------------*/

/* NULL data pointer */
#if !defined NULL
#define NULL                    ( (void *) 0 )
#endif

/****************************************************************************/
/***        Type Definitions                                              ***/
/****************************************************************************/

typedef uint8_t                 bool_t;

/****************************************************************************/
/***        Exported Functions                                            ***/
/****************************************************************************/

#endif // COMMON_H

/****************************************************************************/
/***        END OF FILE                                                   ***/
/****************************************************************************/


orlaco.h
#ifndef ORLACO_H
#define ORLACO_H

/****************************************************************************/
/***        Include files                                                 ***/
/****************************************************************************/

#include <stdint.h>
#include <stdlib.h>

// https://stackoverflow.com/questions/28027937/cross-platform-sockets

#ifdef _WIN32
    /* See http://stackoverflow.com/questions/12765743/getaddrinfo-on-win32 */
    #ifndef _WIN32_WINNT
        #define _WIN32_WINNT 0x0501  /* Windows XP. */
    #endif
    #include <winsock2.h>
    #include <Ws2tcpip.h>
#else
    /* Assume that any non-Windows platform uses POSIX-style sockets instead. */
    #include <sys/socket.h>
    #include <arpa/inet.h>
    #include <netdb.h>  /* Needed for getaddrinfo() and freeaddrinfo() */
    #include <unistd.h> /* Needed for close() */
#endif

#include "common.h"

/****************************************************************************/
/***        Macro Definitions                                             ***/
/****************************************************************************/

#define ORLACO_DEFAULT_PORT             17215
#define ORLACO_MAX_CAMERAS              10
#define ORLACO_MAX_REGISTERS            100
#define ORLACO_MAX_SD_SERVICES          10
#define ORLACO_MAX_SD_OPTIONS           2
#define ORLACO_MAX_RESPONSE_TIME_MS     5000

#define ORLACO_SD_OPTION_LENGTH         0x10

#define ORLACO_BUFFER_LENGTH            1500
#define ORLACO_NUM_REGIONS_OF_INTEREST  11

#ifndef TRUE
#define TRUE                            (1)
#endif

#ifndef FALSE
#define FALSE                           (0)
#endif

/****************************************************************************/
/***        Type Definitions                                              ***/
/****************************************************************************/

typedef enum {
    E_ORLACO_VERBOSITY_ERRORS_ONLY = 0,
    E_ORLACO_VERBOSITY_INFO,
    E_ORLACO_VERBOSITY_DEBUG,
} ORLACO_eVerbosityLevel;

typedef enum {
    E_ORLACO_RETURN_CODE_OK                            = 0x00, // No error occurred
    E_ORLACO_RETURN_CODE_NOT_OK                        = 0x01, // An unspecified error occurred
    E_ORLACO_RETURN_CODE_UNKNOWN_SERVICE               = 0x02, // The requested Service ID is unknown.
    E_ORLACO_RETURN_CODE_UNKNOWN_METHOD                = 0x03, // The requested Method ID is unknown. Service ID is known.
    E_ORLACO_RETURN_CODE_NOT_READY                     = 0x04, // Service ID and Method ID are known. Application not running.
    E_ORLACO_RETURN_CODE_NOT_REACHABLE                 = 0x05, // System running the service is not reachable (internal error code only).
    E_ORLACO_RETURN_CODE_TIMEOUT                       = 0x06, // A timeout occurred (internal error code only).
    E_ORLACO_RETURN_CODE_WRONG_PROTOCOL_VERSION        = 0x07, // Version of SOME/IP protocol not supported
    E_ORLACO_RETURN_CODE_WRONG_INTERFACE_VERSION       = 0x08, // Interface version mismatch
    E_ORLACO_RETURN_CODE_MALFORMED_MESSAGE             = 0x09, // Deserialization error, so that payload cannot be deserialized.
    E_ORLACO_RETURN_CODE_WRONG_MESSAGE_TYPE            = 0x0a, // An unexpected message type was received (e.g. REQUEST_NO_RETURN for a method defined as REQUEST.)

    E_ORLACO_RETURN_CODE_LOCKED_BY_FOREIGN_INSTANCE    = 0x20, // Camera service is already locked by another client
    E_ORLACO_RETURN_CODE_LOCK_EXPIRED                  = 0x21, // The camera lock has expired
    E_ORLACO_RETURN_CODE_NOT_LOCKED                    = 0x22, // Camera is not locked
    E_ORLACO_RETURN_CODE_INVALID_PS_ENTRY              = 0x24, // The requested PSE ID is unknown
    E_ORLACO_RETURN_CODE_INVALID_PS_OPERATION          = 0x25, // The requested PSE operation is not allowed, e.g. store on a RO PSE
    E_ORLACO_RETURN_CODE_INVALID_PS_DATA               = 0x26, // The PSE contains a CRC16 error
    E_ORLACO_RETURN_CODE_NO_MORE_SPACE                 = 0x27, // No more space available to store the PSE
    E_ORLACO_RETURN_CODE_INVALID_ROI_INDEX             = 0x30, // The requested ROI is out of range
    E_ORLACO_RETURN_CODE_INVALID_ROI_NUMBER            = 0x31, // The requested number of ROIs is out of range, defined by sDatasheet.numOfRegionOfInterest
    E_ORLACO_RETURN_CODE_INVALID_VIDEO_FORMAT          = 0x32, // Invalid value in video format
    E_ORLACO_RETURN_CODE_INVALID_HISTOGRAM_FORMAT      = 0x33, // Invalid value in histogram format
    E_ORLACO_RETURN_CODE_INVALID_CONTROL_INDEX         = 0x35, // The requested camControlIndex is out of range or not supported by the camera
    E_ORLACO_RETURN_CODE_INVALID_CONTROL_MODE          = 0x36, // The requested control mode is not supported by the camera
    E_ORLACO_RETURN_CODE_INVALID_CONTROL_VALUE         = 0x37, // The requested control value is out of the range, defined in sCamControl
    E_ORLACO_RETURN_CODE_INVALID_REGISTER_ADDRESS      = 0x38, // The register address is not supported by the imager
    E_ORLACO_RETURN_CODE_INVALID_REGISTER_VALUE        = 0x39, // The value for the given register address is not supported by the imager
    E_ORLACO_RETURN_CODE_INVALID_REGISTER_OPERATION    = 0x3A
} ORLACO_teReturnCode;


// Must make sure these are in the same order as the register addresses
typedef enum {
    E_ORLACO_REGISTER_INDEX_LED_MODE                   = 0,
    E_ORLACO_REGISTER_INDEX_STREAM_PROTOCOL,
    E_ORLACO_REGISTER_INDEX_STATIC_IP_ADDRESS_0,
    E_ORLACO_REGISTER_INDEX_STATIC_IP_ADDRESS_1,
    E_ORLACO_REGISTER_INDEX_STATIC_IP_ADDRESS_2,
    E_ORLACO_REGISTER_INDEX_STATIC_IP_ADDRESS_3,
    E_ORLACO_REGISTER_INDEX_STATIC_NETWORK_MASK_0,
    E_ORLACO_REGISTER_INDEX_STATIC_NETWORK_MASK_1,
    E_ORLACO_REGISTER_INDEX_STATIC_NETWORK_MASK_2,
    E_ORLACO_REGISTER_INDEX_STATIC_NETWORK_MASK_3,
    E_ORLACO_REGISTER_INDEX_MAC_ADDRESS_0,
    E_ORLACO_REGISTER_INDEX_MAC_ADDRESS_1,
    E_ORLACO_REGISTER_INDEX_MAC_ADDRESS_2,
    E_ORLACO_REGISTER_INDEX_MAC_ADDRESS_3,
    E_ORLACO_REGISTER_INDEX_MAC_ADDRESS_4,
    E_ORLACO_REGISTER_INDEX_MAC_ADDRESS_5,
    E_ORLACO_REGISTER_INDEX_VLAN_ID_0,
    E_ORLACO_REGISTER_INDEX_VLAN_ID_1,
    E_ORLACO_REGISTER_INDEX_STREAM_ID_0,
    E_ORLACO_REGISTER_INDEX_STREAM_ID_1,
    E_ORLACO_REGISTER_INDEX_STREAM_ID_2,
    E_ORLACO_REGISTER_INDEX_STREAM_ID_3,
    E_ORLACO_REGISTER_INDEX_STREAM_ID_4,
    E_ORLACO_REGISTER_INDEX_STREAM_ID_5,
    E_ORLACO_REGISTER_INDEX_STREAM_ID_6,
    E_ORLACO_REGISTER_INDEX_STREAM_ID_7,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_IP_ADDRESS_0,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_IP_ADDRESS_1,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_IP_ADDRESS_2,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_IP_ADDRESS_3,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_MAC_ADDRESS_0,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_MAC_ADDRESS_1,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_MAC_ADDRESS_2,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_MAC_ADDRESS_3,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_MAC_ADDRESS_4,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_MAC_ADDRESS_5,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_PORT_0,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_DESTINATION_PORT_1,
    E_ORLACO_REGISTER_INDEX_SELECTED_ROI,
    E_ORLACO_REGISTER_INDEX_NO_STREAM_AT_BOOT,
    E_ORLACO_REGISTER_INDEX_UDP_COMMUNICATION_PORT_0,
    E_ORLACO_REGISTER_INDEX_UDP_COMMUNICATION_PORT_1,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_SOURCE_PORT_0,
    E_ORLACO_REGISTER_INDEX_RTP_STREAM_SOURCE_PORT_1,
    E_ORLACO_REGISTER_INDEX_HDR,
    E_ORLACO_REGISTER_INDEX_OVERLAY,
    E_ORLACO_REGISTER_INDEX_DHCP,
    E_ORLACO_REGISTER_INDEX_WAIT_FOR_MAC,
    E_ORLACO_REGISTER_INDEX_WAIT_FOR_PTP_SYNC,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_0,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_1,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_2,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_3,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_4,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_5,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_6,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_7,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_8,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_9,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_10,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_11,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_12,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_13,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_14,
    E_ORLACO_REGISTER_INDEX_DHCP_HOSTNAME_15,
} ORLACO_teRegisterIndex;

typedef enum {
    E_ORLACO_LED_MODE_OFF                                          = 0x0000,
    E_ORLACO_LED_MODE_AUTO                                         = 0x0001,
    E_ORLACO_LED_MODE_ON                                           = 0x0002,
} ORLACO_teLedMode;

typedef enum {
    E_ORLACO_DHCP_MODE_DISABLED                                    = 0x0000,
    E_ORLACO_DHCP_MODE_ENABLED                                     = 0x0001,
} ORLACO_teDHCPMode;

typedef enum {
    E_ORLACO_VIDEO_COMPRESSION_MODE_NONE                           = 0x00,
    E_ORLACO_VIDEO_COMPRESSION_MODE_JPEG                           = 0x01,
    E_ORLACO_VIDEO_COMPRESSION_MODE_H264                           = 0x02,
} ORLACO_teVideoCompressionMode;


typedef enum {
    E_ORLACO_CAMERA_MODE_START_CAMERA_SERVICE                       = 0x01, // The camera application will be started
    E_ORLACO_CAMERA_MODE_STOP_CAMERA_SERVICE                        = 0x02, // The camera application will be stopped
    E_ORLACO_CAMERA_MODE_RESTART_CAMERA_SERVICE                     = 0x03, // The camera application will be restarted
    E_ORLACO_CAMERA_MODE_STOP_CAMERA                                = 0x04, // The camera will be stopped (standby mode). Requires a power cycle or wake on LAN to recover
} ORLACO_teCameraMode;

typedef struct {
    uint16_t u16Address;
    uint8_t u8Padding;
    uint8_t u8Value;
    char *pcDescription;
    char *pcHelp;
    bool_t bWrite;
    bool_t bRead;
} ORLACO_tsRegisterValue;


typedef union {
    uint8_t au8IP[4];
    uint32_t u32IP;
} ORLACO_tuIP;


typedef struct {
    uint16_t u16P1X;
    uint16_t u16P1Y;
    uint16_t u16P2X;
    uint16_t u16P2Y;
    uint16_t u16OutputWidth;
    uint16_t u16OutputHeight;
    uint32_t u32MaxBitrate;                         // Specified in Megabits per second
    uint8_t u8FrameRate;
    ORLACO_teVideoCompressionMode eCompressionMode;
    bool_t bWrite;
    bool_t bRead;
} ORLACO_tsRegionOfInterest;


typedef struct {
    uint8_t u8Type;
    uint8_t u8Index1stOptions;
    uint8_t u8Index2ndOptions;
    uint8_t u8NumberOfOptions;
    uint16_t u16ServiceID;
    uint16_t u16InstanceID;
    uint8_t u8MajorVersion;
    uint32_t u24TTL;
    uint32_t u32MinorVersion;
} ORLACO_tsServiceDiscoveryServiceEntry;


typedef struct {
    ORLACO_tuIP uIP;
    ORLACO_tsServiceDiscoveryServiceEntry sDiscoveryServiceEntry;
} ORLACO_tsCamera;

#ifdef _WIN32
    typedef unsigned int UDPSOCKET;
#else
    typedef int UDPSOCKET;
#endif

typedef struct {
    ORLACO_eVerbosityLevel eVerbosity;
    struct sockaddr_in fdServer;
    struct sockaddr_in fdUnicast;
    struct sockaddr_in fdBroadcast;
    UDPSOCKET Socket;
    uint16_t u16DstPort;
    uint16_t u16ServiceID;
    uint16_t u16ClientID;
    uint16_t u16SessionID;
    uint16_t u16ResponseTimer;
    ORLACO_teCameraMode eCameraMode;
    uint16_t u16NumRegisters;
    ORLACO_tsRegisterValue *psRegisters;
    uint16_t u16NumRegionsOfInterest;
    ORLACO_tsRegionOfInterest *psRegionsOfInterest;
    uint16_t u16NumCameras;
    ORLACO_tsCamera *psCameras;
} ORLACO_tsInstance;

/****************************************************************************/
/***        Exported Functions                                            ***/
/****************************************************************************/

bool_t ORLACO_bInit(ORLACO_tsInstance *psInstance, char *pcUnicastIP, char *pcBroadcastIP, uint16_t u16DstPort);
void ORLACO_vDeInit(ORLACO_tsInstance *psInstance);
void ORLACO_vSetVerbosity(ORLACO_tsInstance *psInstance, ORLACO_eVerbosityLevel eVerbosityLevel);
bool_t ORLACO_bSetBroadcastIP(ORLACO_tsInstance *psInstance, char *pcIpAddress, uint16_t u16DstPort);
bool_t ORLACO_bSetUnicastIP(ORLACO_tsInstance *psInstance, char *pcIpAddress, uint16_t u16DstPort);

// bool_t ORLACO_bBufferTest(ORLACO_tsInstance *psInstance);
bool_t ORLACO_bDiscover(ORLACO_tsInstance *psInstance);
bool_t ORLACO_bSetCamExclusive(ORLACO_tsInstance *psInstance, uint32_t u32ExclusiveTime);
bool_t ORLACO_bEraseCamExclusive(ORLACO_tsInstance *psInstance);
bool_t ORLACO_bSetCamMode(ORLACO_tsInstance *psInstance, ORLACO_teCameraMode eMode);
bool_t ORLACO_bGetRegisters(ORLACO_tsInstance *psInstance);
bool_t ORLACO_bSetRegisters(ORLACO_tsInstance *psInstance);
bool_t ORLACO_bGetAllRegisters(ORLACO_tsInstance *psInstance);
bool_t ORLACO_bGetRegionOfInterest(ORLACO_tsInstance *psInstance, uint32_t u32RegionOfInterest, ORLACO_tsRegionOfInterest *psRegionOfInterest);
bool_t ORLACO_bGetRegionsOfInterest(ORLACO_tsInstance *psInstance);
bool_t ORLACO_bSetRegionOfInterest(ORLACO_tsInstance *psInstance, uint32_t u32RegionOfInterestIndex, ORLACO_tsRegionOfInterest *psRegionOfInterest);
bool_t ORLACO_bSetRegionsOfInterest(ORLACO_tsInstance *psInstance);
bool_t ORLACO_bSubscribeRoiVideo(ORLACO_tsInstance *psInstance, uint32_t u32RegionOfInterest);


#endif // ORLACO_H

/****************************************************************************/
/***        END OF FILE                                                   ***/
/****************************************************************************/
